bolt-contracts/src/contracts/BoltChallenger.sol

```diff
@@ -0,0 +1,357 @@
+        Challenge storage challenge = challenges[challengeID];
+
+        if (challenge.status != ChallengeStatus.Open) {
+            revert ChallengeAlreadyResolved();
+        }
+
+        if (challenge.openedAt + MAX_CHALLENGE_DURATION >= Time.timestamp()) {
+            revert ChallengeNotExpired();
+        }
+
+        // If the challenge has expired without being resolved, it is considered lost.
+        challenge.status = ChallengeStatus.Lost;
+        _transferFullBond(challenge.challenger);
+        emit ChallengeLost(challengeID);
+    }
+
+    /// @notice Resolve a challenge by providing proofs of the inclusion of the committed transaction.
+    /// @dev Challenges are DEFENDED if the resolver successfully defends the inclusion of the transaction,
+    /// and LOST if the challenger successfully demonstrates that the inclusion commitment was breached or
+    /// enough time has passed without proper resolution.
+    ///
+    /// q: should we also have a commit-reveal scheme for resolutions to avoid frontrunning to steal bonds?
+    function _resolve(bytes32 challengeID, bytes32 trustedBlockHash, Proof calldata proof) internal {
+        // The challenge is assumed to exist at this point, so we can safely access it.
+        Challenge storage challenge = challenges[challengeID];
+
+        if (challenge.status != ChallengeStatus.Open) {
+            revert ChallengeAlreadyResolved();
+        }
+
+        if (challenge.openedAt + MAX_CHALLENGE_DURATION < Time.timestamp()) {
+            // If the challenge has expired without being resolved, it is considered lost.
+            challenge.status = ChallengeStatus.Lost;
+            _transferFullBond(challenge.challenger);
+            emit ChallengeLost(challengeID);
+            return;
+        }
+
+        // Verify the validity of the header against the trusted block hash.
+        if (keccak256(proof.blockHeaderRLP) != trustedBlockHash) {
+            revert InvalidBlockHash();
+        }
+
+        // Decode the RLP-encoded block header fields
+        BlockHeaderData memory blockHeader = _decodeBlockHeaderRLP(proof.blockHeaderRLP);
+
+        // Decode the committed raw signed transaction. Its sender will be the account to prove existence of.
+        TransactionDecoder.Transaction memory decodedTx = challenge.commitment.signedTx.decodeEnveloped();
+
+        // Decode the account fields by checking the account proof against the state root of the block header.
+        // The key in the account trie is the account pubkey (address) which we can recover from the signed tx.
+        (bool accountExists, bytes memory accountRLP) = SecureMerkleTrie.get(
+            abi.encodePacked(decodedTx.recoverSender()), proof.accountMerkleProof, blockHeader.stateRoot
+        );
+
+        if (!accountExists) {
+            revert AccountDoesNotExist();
+        }
+
+        AccountData memory account = _decodeAccountRLP(accountRLP);
+
+        if (account.nonce > decodedTx.nonce) {
```

Comment:
I'm not sure about this check:
- Suppose my account before sending the preconf has nonce $n$. As such, the nonce of the tx is $n$ as well. After it has been included my account will have nonce $n+1$. Since the `proof.accountMerkleProof` is performed against the `blockHeader.stateRoot` after all its transaction I'm worried we might always enter this `if` condition is always satisfied
- What if I send multiple preconfirmations for the same slot? According to the reasoning of the previous point, can we encounter some edge cases?

Reply:
Good points! A simple fix is using the previous blocks' state root to do these checks right?